### PR TITLE
Have compilation context info available earlier in the build process

### PR DESCRIPTION
**Pull request by djc**
Eventually, I hope this will allow us to ignore platform-specific dependencies when irrelevant for the current build earlier in the process. This should save on extraneous errors. As is, this seems like it already decreases coupling in the code base.

### CODE REVIEW

File: src/cargo/core/compiler/mod.rs [hunk 1]
```diff
@@ -76,6 +75,22 @@ pub struct BuildConfig {
...
+    pub fn new(host_triple: &str, requested_target: &Option<String>) -> CargoResult<BuildConfig> {
+        if let Some(ref s) = *requested_target {
+            if s.trim().is_empty() {
+                bail!("target was empty")
+            }
+        }
+        Ok(BuildConfig {
+            host_triple: host_triple.to_string(),
+            requested_target: (*requested_target).clone(),
+            jobs: 1,
```
Comment (matklad) [inline src/cargo/core/compiler/mod.rs]: Let's pass in `jobs: Option<u32>`?

Comment (djc) [reply src/cargo/core/compiler/mod.rs]: The reason I didn't do that is that passing `None` explicitly always feels mostly ugly, especially in this situation where there are only two callers, and `1` seems pretty defensible as a default. All the fields of the struct are already being modified, anyway.

Comment (djc) [reply src/cargo/core/compiler/mod.rs]: (But if you insist, sure!)

**Comment by djc** [comment]
So that last one went a little bit crazy. It does seem like this is much clearer about the actual workings, though, in that `cargo_rustc` did actually not represent any command, but had a whole bunch of code related to driving the compiler, which seems to make more sense in `core`.

**Comment by matklad** [comment]
This looks great to me, but let's also ask @alexcrichton about this refacrtoring!

**Comment by matklad** [comment]
It also seems like RLS might need updating after this PR as well: they use cargo as a library, and so are pretty sensitive to internal organization.

**Pull request merged**